content/blog/handling-exceptions-in-specflow/index.md

```diff
@@ -0,0 +1,323 @@
+---
+
+I use Gherkin scenarios to describe the functional specifications of my software and SpecFlow to automate these scenarios as tests. Usually there will be a couple of scenarios describing the happy path of the feature I'm building but also some scenarios concerning failures. Depending on how the application code works, these failures are represented by exceptions. In this post I explain how I handle these exceptions.
+
+### Table of contents
+
+- [Retrieve existing person successfully](#retrieve-existing-person-successfully)
+- [Retrieve unknown person and expect an error](#retrieve-unknown-person-and-expect-an-error)
+- [Expected error was not raised](#expected-error-was-not-raised)
+- [Check for unexpected errors](#check-for-unexpected-errors)
+- [Refactor to reusable code](#refactor-to-reusable-code)
+  - [Refactored PersonSteps class](#refactored-personsteps-class)
+  - [New generic ErrorSteps](#new-generic-errorsteps)
+  - [The ErrorDriver class](#the-errordriver-class)
+- [Conclusion](#conclusion)
+
+### Retrieve existing person successfully
+
+Let's start with the following happy path scenario to retrieve a person.
+
+```gherkin
+Scenario: Retrieve existing person successfully
+
+Given the person 'Buffy Summers' is registered
+When I retrieve the person 'Buffy Summers'
+Then the person 'Buffy Summers' is returned
+```
+
+In the `Given` step we make sure the person exists in our system. We then retrieve the person and verify that the retrieval is successful.
+
+The following `PersonSteps` class implements this scenario.
+
+```csharp
+[Binding]
+class PersonSteps
+{
+    private readonly PersonRepository _people = new PersonRepository();
+    private string _actualName;
+
+    [Given(@"the person '(.*)' is registered")]
+    public void GivenThePersonIsRegistered(string name)
+    {
+        _people.AddPerson(name);
+    }
+        
+    [When(@"I retrieve the person '(.*)'")]
+    public void WhenIRetrieveThePerson(string name)
+    {
+        _actualName = _people.GetPersonByName(name);
+    }
+
+    [Then(@"the person '(.*)' is returned")]
+    public void ThenThePersonIsReturned(string expectedName)
+    {
+        Assert.IsNotNull(_actualName, "No person retrieved");
+        Assert.AreEqual(expectedName, _actualName);
+    }
+}
+```
+
+It uses a simple in-memory `PersonRepository` to store people. The `_actualName` instance field is used to store the person that is retrieved so we can check if the retrieval was successful in the `Then` step. (For demo purposes we only store and retrieve the name of the person.)
+
+And here's the implementation of `PersonRepository`.
+
+```csharp
+class PersonRepository
+{
+    private readonly HashSet<string> _people = new HashSet<string>();
+
+    public void AddPerson(string name)
+    {
+        _people.Add(name);
+    }
+
+    public string GetPersonByName(string name)
+    {
+        if (_people.Contains(name))
```

Comment:
Beetje vreemde voorbeeld, maar ik snap voor de demo dat het niet belangrijk is :P

Reply:
Ik noem het ook in het blog ergens maar een extra comment hier in het code voorbeeld is nog wel handig.